gef-ui/server/controllers/unissued-facilities/facility-end-date/index.ts

```diff
@@ -0,0 +1,223 @@
+import { Response } from 'express';
+import { getDate, getMonth, getYear, isSameDay, parseISO } from 'date-fns';
+import { CustomExpressRequest, isFacilityEndDateEnabledOnGefVersion, parseDealVersion } from '@ukef/dtfs2-common';
+import { isTrueSet, validationErrorHandler } from '../../../utils/helpers';
+import * as api from '../../../services/api';
+import { validateAndParseFacilityEndDate } from '../../../utils/validate-facility-end-date';
+import { asLoggedInUserSession, LoggedInUserSession } from '../../../utils/express-session';
+import { FacilityEndDateViewModel } from '../../../types/view-models/facility-end-date-view-model';
+import { getCoverStartDateOrStartOfToday } from '../../../utils/get-cover-start-date-or-start-of-today';
+
+type FacilityEndDateParams = { dealId: string; facilityId: string };
+type FacilityEndDatePostBody = { 'facility-end-date-day': string; 'facility-end-date-month': string; 'facility-end-date-year': string };
+
+type GetFacilityEndDateRequest = CustomExpressRequest<{ params: FacilityEndDateParams; query: { change: string } }>;
+type PostFacilityEndDateRequest = CustomExpressRequest<{
+  reqBody: FacilityEndDatePostBody;
+  params: FacilityEndDateParams;
+  query: { saveAndReturn: string };
+}>;
+
+const shouldRedirectFromPage = (dealVersion: number | undefined, isUsingFacilityEndDate: unknown) =>
+  !isFacilityEndDateEnabledOnGefVersion(parseDealVersion(dealVersion)) || !isUsingFacilityEndDate;
+
+const getFacilityEndDateViewModel = (facility: Record<string, unknown>, previousPage: string): FacilityEndDateViewModel => {
+  if (typeof facility.dealId !== 'string' || typeof facility._id !== 'string') {
+    throw new Error('Invalid facility ids stored in database');
+  }
+
+  const facilityEndDateViewModel: FacilityEndDateViewModel = {
+    dealId: facility.dealId,
+    facilityId: facility._id,
+    previousPage,
+  };
+
+  if (typeof facility.facilityEndDate === 'string') {
+    const facilityEndDate = parseISO(facility.facilityEndDate);
+    facilityEndDateViewModel.facilityEndDate = {
+      day: String(getDate(facilityEndDate)),
+      month: String(getMonth(facilityEndDate) + 1),
+      year: String(getYear(facilityEndDate)),
+    };
+  }
+
+  return facilityEndDateViewModel;
+};
+
+export const getFacilityEndDateFromUnissuedFacilitiesPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+    session: { userToken },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/about`;
+
+  try {
+    const { details: facility } = (await api.getFacility({ facilityId, userToken })) as { details: Record<string, unknown> };
+    const deal = (await api.getApplication({ dealId, userToken })) as Record<string, unknown> & { version?: number };
```

Comment:
fwiw, and not now, as these two promises can be awaited for together, we could save execution time here by doing a promise all

Reply:
will do in 7219